posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,368 @@
+own dispatching layer for the API to allow users to develop generic code that
+works with arrays from different libraries. Then it would be possible to
+directly reroute to specific array implementations bypassing compiled code
+paths.
+
+<p align="center">
+    <img
+     alt="A schema using SciPy dispatcher."
+     src="/images/2021/11/scipy-ndimage-dispatcher.png">
+    <i><br>A SciPy module with its own "Dispatcher".</i>
+</p>
+
+<!-- ![scipy.ndimage dispatch schema](https://drive.google.com/uc?export=view&id=1-ZoVxenDufJuUwcONV_mFY0NFNY3wOCq) -->
+
+## Goals, wishes, and constraints
+
+Goal: separate the interface from the implementation and let the dispatching
+system help users and library authors write generic code that works with
+different kinds of arrays.
+
+The new dispatching system implemented for SciPy and scikits should:
+
+* make the API extendable and add the possibility to support alternative array
+  libraries, not just NumPy
+* provide a way to select backend for the same array type (example: `scipy.fft`
+  + alternative CPU FFT libraries, scikit-learn + [Intel optimized
+  implementation](https://github.com/intel/scikit-learn-intelex))
+* be able to extend function signature in a backward-compatible way on the base
+  library side (all alternative implementations should not break after the
+  change)
+* have as low overhead as possible.
+
+When designing API override systems, there are many possible design choices.
+Three major axes of design decisions in the context of NumPy API overrides are
+outlined in [the appendix of NumPy Enhancement Proposal (NEP)
+37](https://numpy.org/neps/nep-0037-array-module.html#appendix-design-choices-for-api-overrides);
+they are also applicable outside of NumPy:
+
+* Opt-in vs. opt-out for users
+* Explicit vs. implicit choice of implementation
+* Local vs. non-local vs. global control.
+
+## A concrete design proposal
+
+In the figure below, we highlight the parts of the design for implementing the
+CuPy backend in SciPy, scikit-learn, scikit-image and other core projects which
+now rely only on NumPy. We are going to use two dispatching mechanisms: one for
+the modules with compiled code portions, and one for the pure Python functions.
+The first dispatching mechanism is based on the [`uarray`
+project](https://uarray.org/), which is a backend dispatcher that allows us to
+choose the relevant function implementation at runtime. The second dispatching
+mechanism is based on the ``__array_namespace__`` method from the Array API
+standard, which is a method that allows us to get the Array API implementation
+module specific to CuPy (or any other array library) at runtime. The
+``__array_namespace__`` will be available in NumPy 1.22 and CuPy v10.0.
+
+The dispatch using ``__array_namespace__`` method is:
+
+* opt-in for users
+  The new dispatching works only when users choose to use the new Array API
+  compatible modules.
```

Comment:
This doesn't seem like a good idea. The opt-in/out approach should work the same for pure Python and compiled code, because that is an implementation detail. This will need discussion on Discourse.

Reply:
I suggest we write it as "One of the key questions to address in this design is whether the behavior should be opt-int or opt-out for users. And if it's enabled by default by a library (and hence opt-out for users, or not even giving users a choice), is it the library which implements the n-D array object that controls this enabling or is it individual array-consuming libraries?"

This was one of the key problems with `__array_function__`. It is a hard question, so I'd like to write it as an actual question like above, and then propose our answer to to it. And reference the uarray auto-registration issue.

So the question then is: what is our proposed answer? Neither of the two extremes seem appealing:
- all opt-out we tried with `__array_function__`, was a blocker for e.g. JAX and PyTorch
- a very granular opt-in per (e.g.) SciPy submodule will be too cumbersome

So the best options seem to be:
- opt-in per package, or
- a global opt-in per array library

Reply:
I agree that it doesn't feel right to have different enabling/registration mechanisms for pure Python or compiled code. It would be confusing and difficult to remember what to use when. I agree that writing it as a question is better.

In the case of Array API individual array-consuming libraries need to choose how they handle inputs both with and without the  `__array_namespace__` method. Using `xp = input.__array_namespace__()` instead of `np = numpy` is not the only possible choice to implement support for Array API compliant libraries. Checking for `__array_namespace__` could be put inside an if statement using global flag, or a context manager.

Maybe the backend dispatching using `uarray` should be enabled only for array objects with `__array_namespace__` method. For other arrays then fall back to the generic implementation coercing to NumPy. Should the context manager (`with uarray.set_backend(...)`) be allowed in this case possibly ignoring absence of `__array_namespace__`?

I think we should encourage auto-registration of `uarray` backends upon import, but the dispatching itself should not be turned on by default on the array-consuming library side. The same switch should control enabling `uarray` dispatching and use of `__array_namespace__` simultaneously to have consistent behavior.

Reply:
> Checking for __array_namespace__ could be put inside an if statement using global flag, or a context manager.

Yes, that's a good point.

> I think we should encourage auto-registration of `uarray` backends upon import, but the dispatching itself should not be turned on by default on the array-consuming library side. The same switch should control enabling `uarray` dispatching and use of `__array_namespace__` simultaneously to have consistent behavior.

I think I like this idea.

Reply:
I removed the current characterization of `uarray` and `__array_namespace__` as it's up to an upcoming discussion to decide what design options to pursue.
I added the proposed question to the text.

Reply:
I like how this reads now. Thanks!